qiskit/qobj/qasm_qobj.py

```diff
@@ -0,0 +1,778 @@
+        """
+        return self._data
+
+    @classmethod
+    def from_dict(cls, data):
+        """Create a new QasmQobjInstruction object from a dictionary.
+
+        Args:
+            data (dict): A dictionary for the experiment config
+
+        Returns:
+            QasmQobjInstruction: The object from the input dictionary.
+        """
+        name = data.pop('name')
+        return cls(name, **data)
+
+    def __eq__(self, other):
+        return_val = False
+        if isinstance(other, QasmQobjInstruction):
+            if self.to_dict() == other.to_dict():
+                return_val = True
+        return return_val
+
+
+class QasmQobjExperiment:
+    """A QASM Qobj Experiment.
+
+    Each instance of this class is used to represent a QASM experiment as
+    part of a larger QASM qobj.
+    """
+
+    def __init__(self, config=None, header=None, instructions=None):
+        """Instatiate a QasmQobjExperiment.
+
+        Args:
+            config (QasmQobjExperimentConfig): A config object for the experiment
+            header (QasmQobjExperimentHeader): A header object for the experiment
+            instructions (list): A list of :class:`QasmQobjInstruction` objects
+        """
+        super(QasmQobjExperiment, self).__init__()
+        self.config = config or QasmQobjExperimentConfig()
+        self.header = header or QasmQobjExperimentHeader()
+        self.instructions = instructions or []
+
+    def to_dict(self):
+        """Return a dictionary format representation of the Experiment.
+
+        Returns:
+            dict: The dictionary form of the QasmQObjExperiment.
+        """
+        out_dict = {
+            'config': self.config.to_dict(),
+            'header': self.header.to_dict(),
+            'instructions': [x.to_dict() for x in self.instructions]
+        }
+        return out_dict
+
+    @classmethod
+    def from_dict(cls, data):
+        """Create a new QasmQobjExperiment object from a dictionary.
+
+        Args:
+            data (dict): A dictionary for the experiment config
+
+        Returns:
+            QasmQobjExperiment: The object from the input dictionary.
+        """
+        config = None
+        if 'config' in data:
+            config = QasmQobjExperimentConfig.from_dict(data.pop('config'))
+        header = None
+        if 'header' in data:
+            header = QasmQobjExperimentHeader.from_dict(data.pop('header'))
+        instructions = None
+        if 'instructions' in data:
+            instructions = [
+                QasmQobjInstruction.from_dict(
+                    inst) for inst in data.pop('instructions')]
+        return cls(config, header, instructions)
+
+    def __eq__(self, other):
+        return_val = False
+        if isinstance(other, QasmQobjExperiment):
+            if self.to_dict() == other.to_dict():
+                return_val = True
+        return return_val
+
+
+class QasmQobjConfig:
+    """A configuration for a QASM Qobj."""
+
+    _data = {}
```

Comment:
Why?

Reply:
This is a weird artifact of enabling `__getattr__` and `__setattr__` around a dict like is needed for backwards compat here. If `_data` is not defined at the class level initially you end up with an infinite recursion trying to access an undefined attribute in getattr. It might not be 100% necessary on this class (it was mostly an issue on the Header and Experiment classes with a `__setattr__` defined. I added it to all the classes that allowed arbitrary fields when I was debugging that failure mode when writing this.